src/main/java/appeng/util/helpers/ItemComparisonHelper.java

```diff
@@ -0,0 +1,443 @@
+				}
+				else
+				{
+					final Item ai = a.getItem();
+					final Item bi = b.getItem();
+
+					final float percentDamagedOfA = 1.0f - (float) ai.getDurabilityForDisplay( a );
+					final float percentDamagedOfB = 1.0f - (float) bi.getDurabilityForDisplay( b );
+
+					return ( percentDamagedOfA > mode.breakPoint ) == ( percentDamagedOfB > mode.breakPoint );
+				}
+			}
+			catch( final Throwable e )
+			{
+				if( mode == FuzzyMode.IGNORE_ALL )
+				{
+					return true;
+				}
+				else if( mode == FuzzyMode.PERCENT_99 )
+				{
+					return ( a.getItemDamage() > 1 ) == ( b.getItemDamage() > 1 );
+				}
+				else
+				{
+					final float percentDamagedOfA = (float) a.getItemDamage() / (float) a.getMaxDamage();
+					final float percentDamagedOfB = (float) b.getItemDamage() / (float) b.getMaxDamage();
+
+					return ( percentDamagedOfA > mode.breakPoint ) == ( percentDamagedOfB > mode.breakPoint );
+				}
+			}
+		}
+
+		final OreReference aOR = OreHelper.INSTANCE.isOre( a );
+		final OreReference bOR = OreHelper.INSTANCE.isOre( b );
+
+		if( OreHelper.INSTANCE.sameOre( aOR, bOR ) )
+		{
+			return true;
+		}
+
+		/*
+		 * // test ore dictionary.. int OreID = getOreID( a ); if ( OreID != -1 ) return OreID == getOreID( b );
+		 * if ( Mode != FuzzyMode.IGNORE_ALL ) { if ( a.hasTagCompound() && !isShared( a.getTagCompound() ) ) { a =
+		 * Platform.getSharedItemStack( AEItemStack.create( a ) ); }
+		 * if ( b.hasTagCompound() && !isShared( b.getTagCompound() ) ) { b = Platform.getSharedItemStack(
+		 * AEItemStack.create( b ) ); }
+		 * // test regular items with damage values and what not... if ( isShared( a.getTagCompound() ) && isShared(
+		 * b.getTagCompound() ) && a.itemID == b.itemID ) { return ((AppEngSharedNBTTagCompound)
+		 * a.getTagCompound()).compareFuzzyWithRegistry( (AppEngSharedNBTTagCompound) b.getTagCompound() ); } }
+		 */
+
+		return a.isItemEqual( b );
+	}
+
+	/**
+	 * recursive test for NBT Equality, this was faster then trying to compare / generate hashes, its also more reliable
+	 * then the vanilla version which likes to fail when NBT Compound data changes order, it is pretty expensive
+	 * performance wise, so try an use shared tag compounds as long as the system remains in AE.
+	 */
+	public boolean isNbtTagEqual( final NBTBase left, final NBTBase right )
+	{
+		// same type?
+		final byte id = left.getId();
+		if( id == right.getId() )
+		{
+			switch( id )
+			{
+				case 10:
+				{
+					final NBTTagCompound ctA = (NBTTagCompound) left;
+					final NBTTagCompound ctB = (NBTTagCompound) right;
+
+					final Set<String> cA = ctA.getKeySet();
+					final Set<String> cB = ctB.getKeySet();
+
+					if( cA.size() != cB.size() )
+					{
+						return false;
+					}
+
+					for( final String name : cA )
+					{
+						final NBTBase tag = ctA.getTag( name );
+						final NBTBase aTag = ctB.getTag( name );
+						if( aTag == null )
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Redundant nullcheck of aTag, which is known to be non-null in appeng.util.helpers.ItemComparisonHelper.isNbtTagEqual(NBTBase, NBTBase) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=findbugs%3ARCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE)